### ISSUE TITLE
Research feasibility of Command#affectsContent

### ISSUE THREAD

**Issue by Reinmar** [ckeditor/ckeditor5#10493]
## Provide a description of the task

Research required to continue #10308.

Notes from a refinement meeting:

*   Prototype this solution.
*   Review the value of `affectsContent` in 73 commands (all that we have now) and set to `false` in these which don't affect the content.
*   Check the aforementioned features. Is this property enough? Or are there more semantically distinctive cases (more props needed)?
*   Search for places where we override the `forceDisabled()` calls with `clearForceDisabled()`
    *   Clean up this code.
    *   Scan for code that could be removed now (hardcoded re-enabling like in FindCommand, hardcoded command names; scan for `clearForceDisabled()`).
*   Are there some commands that do not affect the content but should be disabled in one of the above-mentioned cases (or other cases)?
    *   DONK: comments only mode. Comment commands should be enabled, despite the fact that `affectsContent` would be true.
    *   Still, it can make a decision based on `affectsContent` in [https://github.com/ckeditor/ckeditor5/blob/1fc2435/external/collaboration-features/packages/ckeditor5-comments/src/commentsonly.js#L87](https://github.com/ckeditor/ckeditor5/blob/1fc2435/external/collaboration-features/packages/ckeditor5-comments/src/commentsonly.js#L87)
*   Check whether this will resolve cksource/ckeditor5-internal#796 and similar situations that we had.
*   Give it a thought: Is there a potential command that may need to have this property dynamic (change in different situations).
*   Different direction: Introduce a concept of modes and have `disableInModes[]` property.
    *   Then, a specific command would need to be aware of all modes. E.g. a Bold feature would need to be aware of the Comments feature.
    *   We want to avoid coupling features like that. Hence, a more abstract, and yet fully descriptive property like `affectsContent` is better.
*   Note: In MultiCommand, if any of the registered commands has `affectsContent`, then MultiCommand should too.
*   tl;dr:
    *   Check feasibility (does this property resolve the currently existing cases)
    *   Check the amount of needed changes to reduce this debt

**Comment by dawidurbanski** [comment]
The solution is prototyped (changes: [cke5](https://github.com/ckeditor/ckeditor5/compare/ck/10493-affects-content), [cf](https://github.com/cksource/collaboration-features/compare/ck/10493-affects-content), [internal](https://github.com/cksource/ckeditor5-internal/compare/ck/10493-affects-content)). Here are my thoughts on this after making some research.

First of all, as discussed we should reconsider naming this `affectsData` because `content` word is bit vague. It also follows `Marker` property name as mentioned [here](https://github.com/ckeditor/ckeditor5/issues/10308#issuecomment-893240102).

First of all I created this lookup table and gathered all the data in one place: (internal) https://docs.google.com/spreadsheets/d/1JfNMWffqVP14Mfn4n_JZQpASn2yvrYkD8CWKB_RxkOA/edit#gid=0

<img width="1275" alt="Screenshot 2021-10-04 at 11 55 45" src="https://user-images.githubusercontent.com/12466568/135831323-0cd47adf-f3ee-4542-800f-cdebd08c5e84.png">

### Results can be summarised as such:

- #10636. 
✅  #affectsContent solves this problem
- Replace does not work in Track changes: 
❌ #affectsContent does not solve this problem, because there is simply no integration between these two yet.
- https://github.com/cksource/ckeditor5-internal/issues/1021
✅  #affectsContent solves this problem.
- https://github.com/cksource/ckeditor5-internal/issues/796
✅  #affectsContent solves this problem.
- https://github.com/cksource/collaboration-features/issues/4229
❌ #affectsContent does not solve this problem, because TC has restricted editing integration explicitly allowing these two together (which makes no sense, see the ticket).
- Adding comments threads is enabled in restricted editing.
❌ #affectsContent does not solve this problem, because Comments has restricted editing integration explicitly allowing these two together (which makes no sense, see the ticket from previous point).
- https://github.com/cksource/collaboration-features/issues/4230
❌ #affectsContent does not solve this problem, because in revision history viewer editor is in read-only mode but restoring should be still enabled. It should still be disabled in the proper read only mode. Has to be fixed by CF team.
- #10634
✅  #affectsContent solves this problem (with small exception in read-only mode, check the comment in the table).
- #10635
✅  #affectsContent solves this problem.

### Other related tickets in one place

- https://github.com/ckeditor/ckeditor5/issues/10308
- https://github.com/cksource/ckeditor5-internal/issues/753 (may be closed IMHO)
- https://github.com/cksource/collaboration-features/issues/4213 (find is covered here, replace not)
- https://github.com/cksource/products-feedback/issues/450 (same as previous one)

### New manual test for convenience

I have created a new manual test combining all editor modes in one test for simpler testing.

It's added on branch `ck/10493-affects-content`.

<img width="1327" alt="Screenshot 2021-10-04 at 12 30 12" src="https://user-images.githubusercontent.com/12466568/135836284-d5f2903e-bbe8-48ac-9a05-5dbc9c1d7d47.png">

To use it:

1. `git checkout ck/10493-affects-content`
2. `mrgit checkout ck/10493-affects-content`
3. `yarn manual --files=ckeditor5 -i external/collaboration-features/dev/presets/staging.js`
4. Open `editor-modes.html`

I also created a separate branch `ck/10493-affects-content-compare` with only this test added to current master branch. This can be used for easy comparison of current master vs new affectsContent POC.

To use it:

1. `git checkout ck/10493-affects-content-compare`
2. `mrgit master`
3. `yarn manual --files=ckeditor5 -i external/collaboration-features/dev/presets/staging.js`
4. Open `editor-modes.html`

### Other notes

> Review the value of affectsContent in 73 commands (all that we have now) and set to false in these which don't affect the content.

It's actually `66` commands in main repo, `10` in CF repo and `2` in internal.

> Check the amount of needed changes to reduce this debt

The update is very trivial since we set `affectsContent` on main `Command` class to true by default, and we only have to update this in couple parts, precisely:

set `this.affectsContent = false` in commands:

- [FindCommand](https://github.com/ckeditor/ckeditor5/blob/master/packages/ckeditor5-find-and-replace/src/findcommand.js)
- [FindNextCommand](https://github.com/ckeditor/ckeditor5/blob/master/packages/ckeditor5-find-and-replace/src/findnextcommand.js)
- [RestrictedEditingModeNavigationCommand](https://github.com/ckeditor/ckeditor5/blob/master/packages/ckeditor5-restricted-editing/src/restrictededitingmodenavigationcommand.js)
- [SelectAllCommand](https://github.com/ckeditor/ckeditor5/blob/master/packages/ckeditor5-select-all/src/selectallcommand.js)
- [SelectColumnCommand](https://github.com/ckeditor/ckeditor5/blob/master/packages/ckeditor5-table/src/commands/selectcolumncommand.js)
- [SelectRowCommand](https://github.com/ckeditor/ckeditor5/blob/master/packages/ckeditor5-table/src/commands/selectrowcommand.js)

Internal
- ExportPdfCommand (internal)
- ExportWordCommand (internal)

CF
- ShowChangeCommand (collaboration-features)

add `affectsContent` check in integrations:

- [Read-only in Command](https://github.com/ckeditor/ckeditor5/blob/master/packages/ckeditor5-core/src/command.js#L121)
- [Restricted editing](https://github.com/ckeditor/ckeditor5/blob/e0ff106ede45618f6d1e3ad4922290e967eca3fc/packages/ckeditor5-restricted-editing/src/restrictededitingmodeediting.js#L274-L318) This part also should be refactored to not use command names but rather to loop through all editor commands and disable/enable ones that are necessary (similar to other modes).
- Comments-only (CF CommentsOnly plugin)
- Track changes (CF TrackChangesCommand)

> Is this property enough? Or are there more semantically distinctive cases (more props needed)?

I have not found a single use case for any other props at this stage.

> Are there some commands that do not affect the content but should be disabled in one of the above-mentioned cases.

It's up to the feature to decide whether it wants to respect affectsContent prop, and if not to override default behaviour.

> Is there a potential command that may need to have this property dynamic (change in different situations).

I have not found any use case for this. Also, if a command needs to be content affecting and not affecting at the same time, it might be a sign that it does too much and should be two separate commands (like search + replace for example).

> In MultiCommand, if any of the registered commands has affectsContent, then MultiCommand should too.

In this POC I haven't touch `MultiCommand` so it's something to add in final solution. Should be straightforward.

### What's next?

- Clean up TODOs from this POC
- Add comments
- Add tests if necessary
- Refactor restricted editing command enable/disable part.
- Add `MultiCommand` support for `affectsContent`.
- Maybe rename `affectsContent` to `affectsData`.
